### PR TITLE
saving to and loading from single h5

### PR DESCRIPTION
**Pull request by ReyhaneAskari**
@nouiz
This PR creates a single h5 file instead of many numpy files or a single numpy. Follow up to https://github.com/ruotianluo/ImageCaptioning.pytorch/pull/23.

### CODE REVIEW

**Comment by ruotianluo** [comment]
Does it support parallel?

**Comment by ReyhaneAskari** [comment]
This should fix the parallel problem. For couple of days I was looking for the hdf5 parallel version with SWMR, it was very misleading and not supported properly. At the end, this should work without that. Right now it's running. I'll post a benchmark when it was finished.

**Comment by ruotianluo** [comment]
Can you split the latest commit into two commits: formatting and parallel fix?

**Comment by ReyhaneAskari** [comment]
Sure.

File: dataloader.py [hunk 1]
```diff
@@ -30,7 +30,14 @@ def get_vocab(self):
...
         return self.seq_length
 
+    def read_files(self):
+        self.feats_fc = h5py.File(os.path.join(
+            self.opt.input_fc_dir, 'feats_fc.h5'), 'r')
+        self.feats_att = h5py.File(os.path.join(
+            self.opt.input_att_dir, 'feats_att.h5'), 'r')
+
     def get_npy_data(self, ix):
+        self.read_files()
```
Comment (nouiz) [inline dataloader.py]: why this fix the parallel problem?

Also, if self.feats_{fc,att} aren't needed anymore this function call, I think we should delete them.

Comment (ReyhaneAskari) [reply dataloader.py]: The problem was that before, the file was only opened once and then all the threads wanted to read from that. Now each time we need access we load the files and there is no thread blocking.

`self.feats_fc` is used but we are not using `self.att`.

**Comment by ReyhaneAskari** [comment]
So unfortunately, this is 3 times slower than using the single numpy file. :( 
I guess it is mostly due to the many opening files for the multi-thread. Here is the result for 100 epochs:

|      | Single numpy | Single hdf5 |
|------|--------------|-------------|
| real | 0m48.795s    | 2m43.721s   |
| usr  | 0m38.680s    | 0m29.368s   |
| sys  | 0m17.120s    | 0m29.480s   |

**Comment by ruotianluo** [comment]
I guess it would be faster if we only open the h5 files only once(one file per process)

**Comment by ReyhaneAskari** [comment]
Yes! But the thing is that in that case, I keep getting an error that says unable to open object X, bad header. I don't know if that is a problem with my h5py installation or my code. I "guess"  I need h5py  version 7 and HDF5 version 10 that was built with a special configuration flags for parallelism.

 I worked on updating that for 2 days. I did manage to update my HDF5 by building it from source with the special configuration,  but I couldn't make pip or conda use my current version of HDF5 when installing h5py. I'm not even sure if I manage to install the correct setup, I will not have this problem anymore.
The error that I get is similar to the one in [here](https://github.com/h5py/h5py/issues/991). I do have  a minimal example that reproduces the problem I will post it to h5py issues or pytoch's discussions and update you.

**Comment by ReyhaneAskari** [comment]
Here is the results for NFS hdf5 vs local FS hdf5 and numpy hdf5 when I use to train the model:

|                          | hdf5      | numpy     | local hdf5 |
|--------------------------|-----------|-----------|------------|
| avg read data/batch      | 0.032s    | 0.046s    | 0.032s     |
| avg train time/batch     | 0.072s    | 0.072s    | 0.072s     |
| total time for one epoch | 6m50.312s | 7m22.845s | 6m22.177s  |

the previous timing was only for when I loaded the dataset with this script:

```python
import argparse
from dataloader import DataLoader
import multiprocessing

opt = argparse.Namespace(
    batch_size=32,
    input_att_dir='/data/lisa/data/coco_preprocessed/data/cocotalk_att',
    input_fc_dir='/data/lisa/data/coco_preprocessed/data/cocotalk_fc',
    input_json='/data/lisa/data/coco_preprocessed/data/cocotalk.json',
    input_label_h5='/data/lisa/data/coco_preprocessed/data/cocotalk_label.h5',
    seq_per_img=5, use_att=True, train_only=0)
loader = DataLoader(opt)

print multiprocessing.cpu_count()
# Get one minibatch:
i = 0
for i in range(100):
    data = loader.get_batch('train')
    if i % 10 == 0:
        print i
```
I'm not sure why this script gave higher timing for hdf5 vs numpy.

**Comment by ReyhaneAskari** [comment]
You can check [this thread in pytorch discuss](https://discuss.pytorch.org/t/hdf5-multi-threaded-alternative/6189/4). They have the same issue as we have for hdf5 multi-threading and the only solution they have found that I had found too was to open the file each time. Even HDF Helpdesk has commented there and mentioned  "High levels of concurrency is a problem that needs to be addressed soon." 
So I think this can not be improved unless HDF5 fixes the issue. But as you can see the benchmarks in my previous comment, it is still better than one numpy file.

File: scripts/prepro_feats.py [hunk 2]
```diff
@@ -70,33 +68,51 @@ def main(params):
...
+        print('processing %d/%d (%.2f%% done)' % (i, N, i*100.0 / N))
+    file_fc.close()
+    file_att.close()
+  print('writing table of contents')
+  with open(os.path.join(dir_fc, "toc.json"), 'wt') as fc_toc_file:
+    json.dump(fc_toc, fc_toc_file)
+  with open(os.path.join(dir_att, "toc.json"), 'wt') as att_toc_file:
+    json.dump(att_toc, att_toc_file)
+
   print('wrote ', params['output_dir'])
```
Comment (ruotianluo) [inline scripts/prepro_feats.py]: fc_toc and att_toc is no longer needed right?

File: scripts/prepro_feats.py [hunk 3]
```diff
@@ -70,33 +68,51 @@ def main(params):
...
+      I = torch.from_numpy(I.transpose([2,0,1])).cuda()
+      I = Variable(preprocess(I), volatile=True)
+      tmp_fc, tmp_att = my_resnet(I, params['att_size'])
+      # write to hdf5
+
+      d_set_fc = file_fc.create_dataset(str(img['cocoid']), 
+        (2048,), dtype="float")
+      d_set_att = file_att.create_dataset(str(img['cocoid']), 
+        (params['att_size'], params['att_size'], 2048), dtype="float")
+
```
Comment (ruotianluo) [inline scripts/prepro_feats.py]: What about:
d_set_fc = file_fc.create_dataset(str(img['cocoid']), 
        tmp_fc.shape, dtype="float")
d_set_att = file_att.create_dataset(str(img['cocoid']), 
        tmp_att.shape, dtype="float")

**Comment by ruotianluo** [comment]
Would you mind adding a old_to_new script file which merge the old numpy files into the h5 file? like:
python convert_old.py --input_dir data/cocotalk_att --output_h5 data/cocotalk_att.h5
which browses all the npy or npz files and save into a h5 file. (If it's a npz file, the first element in the dict should be extracted)

**Comment by ReyhaneAskari** [comment]
Thanks for the comments. So just to be clear, the script you need will open each numpy file and save its content to an h5 file, right? Whereas, merge the numpy files themselves into a single h5.

File: scripts/prepro_feats.py (hunk 2 above)
Comment (ReyhaneAskari) [reply scripts/prepro_feats.py]: Yes, done.

File: scripts/prepro_feats.py (hunk 3 above)
Comment (ReyhaneAskari) [reply scripts/prepro_feats.py]: No, this is necessary to create the hdf5 file. I created a dataset for each image with it's id.

**Comment by ReyhaneAskari** [comment]
I added another script which takes all the `.npy` and `.npz` files and construct h5 data. Just like regular process but without feeding to resnet. I need to squash the last two commits but the file system was too slow to do so. I'll do it as soon as possible.

**Pull request merged**